File: iosApp/iosApp/Net/ConnectionHolder.swift
import Foundation
import Client

/// Process-scoped singleton holding the live `LunamuxClient` and
/// `WindowSocket`. Mirrors the Android `ConnectionHolder` singleton.
@Observable
final class ConnectionHolder {
    static let shared = ConnectionHolder()

    private(set) var client: Client.LunamuxClient?
    private(set) var windowSocket: Client.WindowSocket?

    /// Whether a connection is fully established: the client is live **and** the
    /// server's initial config has arrived.
    ///
    /// Deliberately distinct from `client != nil`. `connectMulti` publishes the
    /// client *before* it waits for the config, so the approval-pending flow is
    /// observable while the server-side dialog is up — which means `client`
    /// becomes non-nil during a connect that may still be rejected. Navigation
    /// must key on this flag instead: keying on `client != nil` pushed the
    /// sessions screen the moment the socket opened, and a phase-2 rejection
    /// then cleared the client without ever popping that screen, stranding the
    /// user on an empty session list behind the failure alert.
    ///
    /// Read by `HostsView` to decide when to navigate to the sessions screen.
    private(set) var isReady: Bool = false

    private init() {
        client = nil
        windowSocket = nil
    }

    /// Whether the server is showing an approval dialog for this connection.
    private(set) var pendingApproval: Bool = false

    /// Set to `true` by `connect(...)` when the most recent attempt failed
    /// specifically because the server's leaf cert no longer matches the
    /// stored pin (verify-mode mismatch in the shared
    /// `PinnedHttpClientFactory`). Callers branch on this in their `catch`
    /// to show a cert-changed dialog instead of a generic error — Darwin's
    /// `URLSessionDelegate` cancels the challenge with no marker NSError,
    /// so the bridged Swift error is otherwise indistinguishable from a
    /// wrong-port / unreachable-host failure.
    private(set) var lastPinMismatch: Bool = false

    private var approvalObserver: Client.FlowObserver?

    /// Tear down any existing client and create a fresh one for the given URL.
    ///
    /// Retained for the built-in demo path only (see `HostsViewModel.connectDemo`);
    /// real hosts go through `connectMulti`, which walks every candidate
    /// endpoint a paired entry carries. Mirrors Android, where `connect` was
    /// narrowed to the demo for the same reason.
    ///
    /// `pinnedFingerprintHex` selects between the TOFU capture mode (nil,
    /// first-connect to a new host — the leaf cert's SHA-256 is observed
    /// during the handshake and exposed via `client.observedFingerprint`)
    /// and verify mode (non-nil — `URLSessionDelegate` constant-time compares
    /// the leaf against the pin and cancels the challenge on mismatch).
    /// A mismatch additionally sets `lastPinMismatch = true` so the caller's
    /// `catch` can show the cert-changed UX instead of a generic error.
    @MainActor
    func connect(
        serverUrl: Client.ServerUrl,
        authToken: String,
        pinnedFingerprintHex: String? = nil
    ) async throws {
        disconnect()
        pendingApproval = false
        lastPinMismatch = false
        let identity = Self.identity(
            demo: Client.DemoModeKt.isDemoHost(host: serverUrl.host)
        )
        let fresh = Client.LunamuxClientKt.createLunamuxClient(
            serverUrl: serverUrl,
            authToken: authToken,
            identity: identity,
            pinnedFingerprintHex: pinnedFingerprintHex,
            // Pairing tokens only ever ride the `connectMulti` path; this
            // single-endpoint connect is the demo's, which has no server to
            // pair with.
            pairingToken: nil
        )
        let socket = fresh.openWindowSocket()
        // Watch for PendingApproval so we can update the UI while
        // waiting for the server-side dialog to be answered.
        let observer = Client.FlowObserver()
        approvalObserver = observer
        observer.observe(flow: fresh.windowState.pendingApproval) { [weak self] value in
            DispatchQueue.main.async {
                self?.pendingApproval = (value as? Bool) == true
            }
        }
        do {
            // Phase 1: WebSocket handshake — 15 s is plenty.
            try await withTimeout(seconds: 15) {
                try await socket.awaitSessionReady()
            }
            // Phase 2: wait for the first Config envelope.  When approval
            // is pending this can take minutes, so use a generous timeout.
            try await withTimeout(seconds: 300) {
                try await socket.awaitInitialConfig()
            }
            pendingApproval = false
        } catch {
            pendingApproval = false
            observer.clear()
            approvalObserver = nil
            // Sample the side-channel set by the shared
            // PinnedHttpClientFactory.ios before we tear the client down —
            // the StateFlow value stays readable after close(), but reading
            // it here keeps the lifetime obvious.
            lastPinMismatch = (fresh.observedMismatch.value as? String) != nil
            let nsError = error as NSError
            NSLog(
                "[ConnectionHolder] connect failed: domain=%@ code=%ld desc=%@ userInfo=%@",
                nsError.domain, nsError.code,
                nsError.localizedDescription, String(describing: nsError.userInfo)
            )
            try? await socket.close()
            fresh.close()
            throw error
        }
        observer.clear()
        approvalObserver = nil
        self.client = fresh
        self.windowSocket = socket
        // Both phases passed — safe to navigate.
        self.isReady = true
    }

    /// Try every candidate endpoint of a host entry in order and keep the
    /// first that connects. Used by the hosts screen for saved entries — a
    /// QR-paired entry carries every address the server advertised, and even
    /// manually-added ones benefit from the same code path.
    ///
    /// Phase 1 (per candidate, inside the shared `CandidateConnector`):
    /// WebSocket handshake with a 12 s budget. Phase 2 (winner only): wait for
    /// the initial Config envelope, up to 5 min so a server-side approval
    /// dialog can be answered. The client is published before phase 2 so
    /// `pendingApproval` is observable while waiting.
    ///
    /// Mirrors the Android `ConnectionHolder.connectMulti`, down to the
    /// phase-1/phase-2 error split: only an unreachable phase 1 is blamed on
    /// the network.
    ///
    /// - Parameters:
    ///   - addresses: ordered endpoints to try — typically a host entry's
    ///     `addresses` verbatim. See `HostEntryLocal.addresses`.
    ///   - authToken: the device-auth token.
    ///   - pinnedFingerprintHex: TLS pin (verify mode), or `nil` for TOFU
    ///     capture on first connect.
    ///   - pairingToken: one-time QR pairing token, or `nil` outside pairing.
    ///   - onAttempt: invoked with each endpoint just before it is tried, so
    ///     the UI can name what it is waiting on. Passes the endpoint rather
    ///     than a formatted string: the hosts screen matches it against the
    ///     walk to show "address 2 of 3", which a string would force it to
    ///     re-parse.
    /// - Returns: the winning `CandidateConnection`; the caller promotes its
    ///   endpoint to the head of the entry's address list.
    /// - Throws: the shared `ServerUnreachableException` when no address
    ///   answered, `DeviceAuthRejectedException` when the server refused the
    ///   device, or the underlying failure otherwise — all typed by shared code
    ///   and classified for the user by `ConnectFailureCopy`. Pin mismatches
    ///   propagate so the caller can check `lastPinMismatch`.
    @MainActor
    @discardableResult
    func connectMulti(
        addresses: [Client.HostPort],
        authToken: String,
        pinnedFingerprintHex: String? = nil,
        pairingToken: String? = nil,
        onAttempt: @escaping (Client.HostPort) -> Void = { _ in }
    ) async throws -> Client.CandidateConnection {
        disconnect()
        pendingApproval = false
        lastPinMismatch = false

        // Real hosts only — the demo never reaches `connectMulti`.
        let identity = Self.identity(demo: false)

        // Phase 1 — reach an address. A pin mismatch is re-surfaced through
        // `lastPinMismatch` so the UI can show the cert-changed dialog. Every
        // other failure is already typed by the shared connector — an
        // unreachable walk arrives as `ServerUnreachableException` — so it is
        // rethrown untouched rather than re-wrapped here; wrapping it in a
        // Swift type was what forced the hosts screen to re-derive, in Swift,
        // a fact shared code had already established.
        let connection: Client.CandidateConnection
        do {
            connection = try await Client.CandidateConnector.shared.connectFirstReachable(
                endpoints: addresses,
                authToken: authToken,
                identity: identity,
                pinnedFingerprintHex: pinnedFingerprintHex,
                pairingToken: pairingToken,
                perCandidateTimeoutMs: Client.CandidateConnector.shared
                    .DEFAULT_PER_CANDIDATE_TIMEOUT_MS,
                onAttempt: onAttempt
            )
        } catch {
            if Self.isPinMismatch(error) {
                lastPinMismatch = true
            }
            throw error
        }

        // Publish before the config wait so the approval-pending flow is
        // observable while the server-side dialog (if any) is up.
        let observer = Client.FlowObserver()
        approvalObserver = observer
        observer.observe(flow: connection.client.windowState.pendingApproval) { [weak self] value in
            DispatchQueue.main.async {
                self?.pendingApproval = (value as? Bool) == true
            }
        }
        self.client = connection.client
        self.windowSocket = connection.windowSocket

        do {
            try await withTimeout(seconds: 300) {
                try await connection.windowSocket.awaitInitialConfig()
            }
        } catch {
            NSLog("[ConnectionHolder] connectMulti failed awaiting config: %@",
                  error.localizedDescription)
            pendingApproval = false
            observer.clear()
            approvalObserver = nil
            disconnect()
            throw error
        }
        pendingApproval = false
        observer.clear()
        approvalObserver = nil
        // The config landed — this connection is real. Only now may the UI
        // leave the hosts screen.
        isReady = true
        return connection
    }

    /// Whether `error` is the shared connector's pin-mismatch signal.
    ///
    /// Kotlin exceptions crossing into Swift arrive as an `NSError` carrying
    /// the original throwable under `KotlinException`, so the cause-chain scan
    /// that Android runs directly is reachable here too — but only because
    /// `CandidateConnector` wraps the Darwin case in a `PinMismatchException`
    /// first. Darwin's `URLSessionDelegate` cancels the auth challenge with a
    /// generic error, so without that wrapper there would be nothing to match.
    ///
    /// - Parameter error: a failure thrown out of `connectFirstReachable`.
    /// - Returns: `true` when the server's leaf cert failed the pin check.
    private static func isPinMismatch(_ error: Error) -> Bool {
        let nsError = error as NSError
        if let thrown = nsError.userInfo["KotlinException"] as? Client.KotlinThrowable,
           Client.CandidateConnector.shared.isPinMismatch(t: thrown) {
            return true
        }
        // Fallback: the boxed throwable is the documented carrier, but the
        // marker also survives into the bridged description.
        return nsError.localizedDescription.contains("pin-mismatch:")
    }

    @MainActor
    func disconnect() {
        let ws = windowSocket
        isReady = false
        windowSocket = nil
        client?.close()
        client = nil
        Task { try? await ws?.close() }
    }

    /// Self-reported identity for this device.
    ///
    /// Reports type "iOS" so the settings UI can tell this apart from Android
    /// and browser tabs, and the running app version (CFBundleShortVersionString)
    /// so the server can gate newer pane kinds (agent consoles, 1.5+) to
    /// clients able to render them. Both host/IP fields are advisory.
    ///
    /// The host/IP lookups are skipped entirely in demo mode, and not merely
    /// as an optimisation: `ProcessInfo.hostName` resolves the device's own
    /// `.local` name over mDNS, which trips the iOS local-network permission
    /// alert. The demo runs against the in-process `DemoServer` and opens no
    /// socket at all, so prompting for LAN access there asks the user to grant
    /// a permission nothing will use — and it lands before they have any
    /// reason to trust the app. There is also no server on the far end to
    /// report an identity to. Real connects still resolve both: the prompt is
    /// then answered in a context where the app genuinely needs the LAN.
    ///
    /// - Parameter demo: whether this connect targets the built-in demo.
    /// - Returns: the identity to hand to the shared client.
    /// - SeeAlso: `HostsViewModel.connectDemo`
    private static func identity(demo: Bool) -> Client.ClientIdentity {
        let appVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String
        return Client.ClientIdentity(
            type: "iOS",
            hostname: demo ? nil : ProcessInfo.processInfo.hostName,
            selfReportedIp: demo ? nil : Self.firstNonLoopbackIPv4(),
            version: appVersion
        )
    }

    /// Best-effort first non-loopback IPv4 address.
    private static func firstNonLoopbackIPv4() -> String? {
        var ifaddr: UnsafeMutablePointer<ifaddrs>?
        guard getifaddrs(&ifaddr) == 0, let first = ifaddr else { return nil }
        defer { freeifaddrs(first) }

        var cursor: UnsafeMutablePointer<ifaddrs>? = first
        while let ifa = cursor {
            let sa = ifa.pointee.ifa_addr
            if sa?.pointee.sa_family == UInt8(AF_INET) {
                let name = String(cString: ifa.pointee.ifa_name)
                if name != "lo0" {
                    var addr = [CChar](repeating: 0, count: Int(NI_MAXHOST))
                    if getnameinfo(sa, socklen_t(sa!.pointee.sa_len),
                                   &addr, socklen_t(addr.count),
                                   nil, 0, NI_NUMERICHOST) == 0 {
                        return String(cString: addr)
                    }
                }
            }
            cursor = ifa.pointee.ifa_next
        }
        return nil
    }
}

/// Swift-friendly timeout wrapper for async operations.
private func withTimeout<T>(seconds: TimeInterval, operation: @escaping () async throws -> T) async throws -> T {
    try await withThrowingTaskGroup(of: T.self) { group in
        group.addTask { try await operation() }
        group.addTask {
            try await Task.sleep(nanoseconds: UInt64(seconds * 1_000_000_000))
            throw NSError(domain: "ConnectionHolder", code: -1,
                          userInfo: [NSLocalizedDescriptionKey: "Connection timed out"])
        }
        let result = try await group.next()!
        group.cancelAll()
        return result
    }
}

File: androidApp/src/main/kotlin/se/soderbjorn/lunamux/android/net/ConnectionHolder.kt
package se.soderbjorn.lunamux.android.net

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeout
import se.soderbjorn.lunamux.HostPort
import se.soderbjorn.lunamux.android.BuildConfig
import se.soderbjorn.lunamux.client.CandidateConnection
import se.soderbjorn.lunamux.client.CandidateConnector
import se.soderbjorn.lunamux.client.ClientIdentity
import se.soderbjorn.lunamux.client.ServerUrl
import se.soderbjorn.lunamux.client.LunamuxClient
import se.soderbjorn.lunamux.client.WindowSocket
import se.soderbjorn.lunamux.client.demo.isDemoHost
import java.net.Inet4Address
import java.net.InetAddress
import java.net.NetworkInterface

/**
 * Process-scoped holder for the live [LunamuxClient]. The app is tiny
 * enough that a plain singleton is cleaner than pulling in Hilt or Koin.
 * [LunamuxApp] rebuilds this whenever the user commits a new host/port
 * from the Connect screen.
 */
object ConnectionHolder {
    @Volatile
    private var currentClient: LunamuxClient? = null

    @Volatile
    private var currentWindowSocket: WindowSocket? = null

    /**
     * Returns the currently connected [LunamuxClient], or null if disconnected.
     *
     * @return the active client instance, or null.
     */
    fun client(): LunamuxClient? = currentClient

    /**
     * Returns the open [WindowSocket] for the current connection, or null if disconnected.
     *
     * @return the active window socket, or null.
     */
    fun windowSocket(): WindowSocket? = currentWindowSocket

    /**
     * Reflects whether the server is currently showing a device-approval
     * dialog for this connection. UI can observe this to show
     * "Waiting for approval…" instead of a bare spinner.
     */
    val pendingApproval: StateFlow<Boolean>?
        get() = currentClient?.windowState?.pendingApproval

    /**
     * Try every address of a host entry in order and keep the first that
     * connects. Used by the hosts screen for saved entries — QR-paired entries
     * carry several addresses, and even manually-added ones benefit from the
     * same code path.
     *
     * Phase 1 (per address, inside [CandidateConnector]): WebSocket handshake
     * with a 12 s budget. Phase 2 (winner only): wait for the initial Config
     * envelope, up to 5 min so a server-side approval dialog can be answered.
     * The current client is published before phase 2 so [pendingApproval] is
     * observable while waiting.
     *
     * @param addresses ordered endpoints to try — typically a host entry's
     *   [se.soderbjorn.lunamux.client.HostEntry.addresses] verbatim.
     * @param authToken the device-auth token.
     * @param pinnedFingerprintHex TLS pin (verify mode), or `null` for TOFU
     *   capture on first connect.
     * @param pairingToken one-time QR pairing token, or `null` outside the
     *   pairing flow.
     * @param onAttempt invoked with each endpoint just before it is tried, so
     *   the UI can name what it is waiting on. Passes the endpoint rather than
     *   a formatted string: the hosts screen matches it against the walk to
     *   show "address 2 of 3", which a string would force it to re-parse.
     * @return the winning [CandidateConnection]; the caller promotes its
     *   endpoint to the head of the entry's address list.
     * @throws Throwable when every address fails — pin-mismatch failures
     *   take precedence (see [CandidateConnector.connectFirstReachable]).
     */
    suspend fun connectMulti(
        addresses: List<HostPort>,
        authToken: String,
        pinnedFingerprintHex: String? = null,
        pairingToken: String? = null,
        onAttempt: (HostPort) -> Unit = {},
    ): CandidateConnection {
        disconnect()
        Log.i("ConnectionHolder", "connectMulti: ${addresses.size} address(es)")
        // Phase 1 — reach an address. Every failure is already typed by the
        // shared connector: a pin mismatch propagates for the cert-changed
        // dialog, and an unreachable walk arrives as ServerUnreachableException.
        // Nothing is re-wrapped here — doing so was what forced the hosts screen
        // to re-derive, in Android code, a fact shared code had established.
        val connection = CandidateConnector.connectFirstReachable(
            endpoints = addresses,
            authToken = authToken,
            // Real hosts only — the demo never reaches connectMulti.
            identity = androidIdentity(demo = false),
            pinnedFingerprintHex = pinnedFingerprintHex,
            pairingToken = pairingToken,
            onAttempt = onAttempt,
        )
        // Publish before the config wait so the approval-pending flow is
        // observable while the server-side dialog (if any) is up.
        currentClient = connection.client
        currentWindowSocket = connection.windowSocket
        try {
            withTimeout(300_000) {
                connection.windowSocket.awaitInitialConfig()
            }
        } catch (t: Throwable) {
            Log.w("ConnectionHolder", "connectMulti failed awaiting config", t)
            disconnect()
            throw t
        }
        return connection
    }

    /**
     * Self-reported identity for this device. Android devices report type
     * "Android" so the settings UI can tell them apart from iOS and browser
     * tabs, and the running app version so the server can gate newer pane
     * kinds (agent consoles, 1.5+) to clients that can render them. The
     * hostname + first non-loopback IPv4 lookups are best-effort and
     * advisory; a failure just means blanks.
     *
     * Runs on [Dispatchers.IO]: `getLocalHost()` does a reverse-DNS lookup
     * that can block for seconds on a bad resolver, and connect callers run
     * on the main dispatcher — resolving it here keeps that off the UI thread.
     *
     * In demo mode both lookups are skipped: the demo runs against the
     * in-process `DemoServer` and opens no socket, so there is no server to
     * report an identity to and no reason to make the user wait on a resolver
     * that may block for seconds. Mirrors the iOS `ConnectionHolder.identity`,
     * where skipping the same pair of lookups additionally avoids tripping the
     * local-network permission alert.
     *
     * @param demo whether this connect targets the built-in demo.
     * @return the identity to hand to the shared client.
     */
    private suspend fun androidIdentity(demo: Boolean): ClientIdentity = withContext(Dispatchers.IO) {
        ClientIdentity(
            type = "Android",
            hostname = if (demo) null else runCatching { InetAddress.getLocalHost().hostName }.getOrNull(),
            selfReportedIp = if (demo) null else runCatching { firstNonLoopbackIpv4() }.getOrNull(),
            version = BuildConfig.VERSION_NAME,
        )
    }

    /**
     * Tears down any existing client and creates a fresh one for [serverUrl].
     *
     * Retained for the built-in demo path only (see
     * [se.soderbjorn.lunamux.android.ui.HostsScreen]'s demo footer);
     * real hosts go through [connectMulti]. Performs the same two-phase
     * handshake: WebSocket session (15 s), then the first Config envelope
     * (up to 5 min for a device-approval dialog).
     *
     * @param serverUrl the server URL containing host and port.
     * @param authToken the authentication token for this client.
     * @param pinnedFingerprintHex lowercase hex SHA-256 of the server's leaf
     *   cert (verify mode), or `null` to run TOFU capture on first connect.
     *   The caller can read `client().observedFingerprint.value` afterwards
     *   to persist the captured pin.
     * @return the connected [LunamuxClient] instance.
     * @throws Throwable if the connection or handshake fails. A
     *   `javax.net.ssl.SSLHandshakeException` whose cause chain contains a
     *   `CertificateException` starting with `"pin-mismatch:"` indicates the
     *   server's cert no longer matches the stored pin; UI should surface
     *   the cert-changed dialog (see [isPinMismatch]).
     * @see connectMulti
     */
    suspend fun connect(
        serverUrl: ServerUrl,
        authToken: String,
        pinnedFingerprintHex: String? = null,
    ): LunamuxClient {
        disconnect()
        Log.i("ConnectionHolder", "connect: ${serverUrl.wsUrl("/window")}")
        val fresh = LunamuxClient(
            serverUrl = serverUrl,
            authToken = authToken,
            identity = androidIdentity(demo = isDemoHost(serverUrl.host)),
            pinnedFingerprintHex = pinnedFingerprintHex,
        )
        val socket = fresh.openWindowSocket()
        try {
            // Phase 1: WebSocket handshake — 15 s is plenty.
            withTimeout(15_000) {
                socket.awaitSessionReady()
            }
            // Phase 2: wait for the first Config envelope. When device
            // approval is pending, the server-side dialog can take minutes
            // to be answered, so use a generous timeout. The client
            // receives a PendingApproval envelope immediately (observable
            // via [pendingApproval]) so the UI can show feedback.
            withTimeout(300_000) {
                socket.awaitInitialConfig()
            }
        } catch (t: Throwable) {
            Log.w("ConnectionHolder", "connect failed", t)
            runCatching { socket.close() }
            runCatching { fresh.close() }
            throw t
        }
        currentClient = fresh
        currentWindowSocket = socket
        return fresh
    }

    /**
     * Scans all network interfaces for the first non-loopback, non-link-local
     * IPv4 address to use as the advisory self-reported IP in the client identity.
     *
     * @return the IP address string, or null if none found.
     */
    private fun firstNonLoopbackIpv4(): String? {
        val nics = NetworkInterface.getNetworkInterfaces() ?: return null
        for (nic in nics) {
            if (!nic.isUp || nic.isLoopback) continue
            for (addr in nic.inetAddresses) {
                if (addr is Inet4Address && !addr.isLoopbackAddress && !addr.isLinkLocalAddress) {
                    return addr.hostAddress
                }
            }
        }
        return null
    }

    /**
     * App-resume hook: if the `/window` socket has been quiet long enough
     * to be presumed dead (the OS silently kills TCP connections while the
     * phone sleeps; the read loop never errors), force an immediate
     * reconnect so the tree/screens refresh with the server's current
     * state. No-op when disconnected or when the connection is healthy —
     * the server pushes session states every ~3 s, so a healthy socket is
     * never quiet for [WINDOW_STALE_MS].
     *
     * Called from [se.soderbjorn.lunamux.android.ui.LunamuxApp]'s
     * lifecycle observer on every ON_START.
     */
    fun refreshAfterResume() {
        currentWindowSocket?.reconnectIfStale(WINDOW_STALE_MS)
    }

    /** Quiet threshold for [refreshAfterResume]; see the 3 s state poller. */
    private const val WINDOW_STALE_MS = 8_000L

    /**
     * Closes the current window socket and client, resetting both to null.
     * Safe to call even when already disconnected.
     */
    suspend fun disconnect() {
        currentWindowSocket?.close()
        currentWindowSocket = null
        currentClient?.close()
        currentClient = null
    }

    /**
     * Walk a throwable's cause chain looking for the `"pin-mismatch:"` marker
     * thrown by [se.soderbjorn.lunamux.client.createPinnedHttpClient] when
     * the server's leaf cert no longer matches the stored pin. Used by
     * [se.soderbjorn.lunamux.android.ui.HostsScreen] to decide between
     * showing a generic error snackbar and a cert-changed re-pair dialog.
     *
     * The scan itself now lives in the shared [CandidateConnector] (so iOS
     * gets it too); this delegate is kept for existing call sites.
     *
     * @param t the throwable to inspect (typically the failure surfaced from
     *   [connect] / [connectMulti]).
     * @return `true` if any link in the cause chain is a pin-mismatch.
     */
    fun isPinMismatch(t: Throwable): Boolean = CandidateConnector.isPinMismatch(t)
}
